specs/Standards/Wallets/InjectedWallets.md

```diff
@@ -0,0 +1,369 @@
+
+```ts
+await window.near.wallet.disconnect();
+```
+
+##### `signIn`
+
+Add `FunctionCall` access key(s) for one or more accounts. This request should require explicit approval from the user.
+
+```ts
+import { utils } from "near-api-js";
+
+// Retrieve the list of accounts we have visibility of.
+const { accounts } = window.near.wallet;
+
+// Request FunctionCall access to the 'guest-book.testnet' smart contract for each account.
+await window.near.wallet.signIn({
+  permission: {
+    receiverId: "guest-book.testnet",
+    methodNames: [],
+  },
+  accounts: accounts.map(({ accountId }) => {
+    const keyPair = utils.KeyPair.fromRandom("ed25519");
+
+    return {
+      accountId,
+      publicKey: keyPair.getPublicKey()
+    };
+  }),
+});
+```
+
+##### `signOut`
+
+Delete `FunctionCall` access key(s) for one or more accounts. This request should require explicit approval from the user.
+
+```ts
+import { utils, keyStores } from "near-api-js";
+
+// Setup keystore to retrieve locally stored FunctionCall access keys.
+const keystore = new keyStores.BrowserLocalStorageKeyStore();
+
+// Retrieve accounts (assuming already connected) and current network.
+const { network, accounts } = window.near.wallet;
+
+// Remove FunctionCall access (previously granted via signIn) for each account.
+await window.near.wallet.signOut({
+  accounts: await Promise.all(
+    accounts.map(async ({ accountId }) => {
+      const keyPair = await keystore.getKey(network.networkId, accountId);
+
+      return {
+        accountId,
+        publicKey: keyPair.getPublicKey()
+      };
+    })
+  ),
+});
+```
+
+#### Events
+
+##### `accountsChanged`
+
+Triggered whenever accounts are updated (e.g. calling `connect` or `disconnect`).
+
+```ts
+window.near.wallet.on("accountsChanged", ({ accounts }) => {
+  console.log("Accounts Changed", accounts);
+});
+```
+
+##### `networkChanged`
+
+Triggered whenever the wallet changes network.
+
+```ts
+window.near.wallet.on("networkChanged", ({ network }) => {
+  console.log("Network Changed", network);
+});
+```
+
+## Drawbacks
+
+There are cases when a dApp wants to sign a transaction with a signer that's yet to be determined (e.g. "Donate" button). Using the `Transaction` instance from `near-api-js` sadly limits this possibility as it requires the details of a `signerId` and `publicKey` upfront. Below are suggested solutions:
+
+- A workaround that maintains the `Transaction` instance might involve creating transactions with empty an `signerId` and `publicKey` to signal to wallets during validation that a prompt is required to select an account (similar to `connect`) before reconstructing the transaction.
+- Consider an intermediate abstraction between the dApp and wallet to defer some parameters for transactions (or at least make them optional). This gives wallets an opportunity to prompt for missing data and in some cases reduce complexity for dApps as the `nonce` and `blockHash` can be handled internally.
```

Comment:
I agree with this. Could we use something similar to the [`SignAndSendTransactionOptions`](https://github.com/near/near-api-js/blob/master/packages/near-api-js/src/account.ts#L66) type used in [near-api-js](https://github.com/near/near-api-js) ?  This is currently used in `Account` and [`ConnectedWalletAccount`](https://github.com/near/near-api-js/blob/master/packages/near-api-js/src/wallet-account.ts#L298). We could modify it slightly to include an optional `signerId` property:

```typescript
interface TransactionOptions {
    receiverId: string;
    actions: Array<transactions.Action>;
    signerId?: string;
}

interface SignTransactionParams {
  transaction: TransactionOptions;
}

interface SignTransactionsParams {
  transactions: Array<TransactionOptions>;
}

signTransaction(params: SignTransactionParams): Promise<transactions.SignedTransaction>;
signTransactions(params: SignTransactionsParams): Promise<Array<transactions.SignedTransaction>>;
```

This has the added benefit of moving [the work of constructing a transaction](https://docs.near.org/integrator/create-transactions#transaction-requirements) to the wallet side including having to fetch and increment the key nonce and decoding the `blockHash` using `borsh`.

Also, this way the dapp does not have to call `connect` if it just wants a transaction signed by an arbitrary `accountId`.

Reply:
:+1: Making this adjustment as suggested - thanks for the discussion folks!